src/python/platforms/android/emulator.py

```diff
@@ -0,0 +1,126 @@
+
+import os
+import shutil
+import subprocess
+import tempfile
+import time
+
+from google_cloud_utils import storage
+from metrics import logs
+from platforms.android import adb
+from system import archive
+from system import environment
+from system import new_process
+
+_WAIT_SECONDS = 10
+
+_emu_proc = None
+_emu_users = 0
+
+
+class EmulatorError(Exception):
+  """Error for errors handling the Android emulator."""
+
+
+class EmulatorProcess(object):
+  """A EmulatorProcess encapsulates the creation, running, and destruction
+  of Android emulator processes."""
+
+  def __init__(self):
+    self.process_runner = None
+    self.popen = None
+    self.logfile = None
+
+    log_path = os.path.join(tempfile.gettempdir(), 'android-emulator-log')
+    self.logfile = open(log_path, 'wb')
+
+  def create(self):
+    """Configures a emulator process which can subsequently be `run`."""
+    # Download emulator image
+    temp_directory = environment.get_value('BOT_TMPDIR')
+
+    archive_src_path = environment.get_value('ANDROID_EMULATOR_BUCKET_PATH')
+    archive_dst_path = os.path.join(temp_directory, 'emulator_bundle.zip')
+
+    storage.copy_file_from(archive_src_path, archive_dst_path)
+
+    # Extract emulator image.
+    self.emu_path = os.path.join(temp_directory, 'emulator')
+    archive.unpack(archive_dst_path, self.emu_path)
+    os.remove(archive_dst_path)
+
+    # Run emulator
+    script_path = os.path.join(
+        self.emu_path, environment.get_value('ANDROID_EMULATOR_SCRIPT_PATH'))
+
+    self.process_runner = new_process.ProcessRunner(script_path, [])
+
+  def run(self):
+    """Actually runs a emulator, assuming `create` has already been called."""
+    if not self.process_runner:
+      raise EmulatorError('Attempted to `run` emulator before calling `create`')
+
+    devices_before = adb.get_devices()
+    new_device = False
+
+    logs.log('Starting emulator.')
+    self.popen = self.process_runner.run(
+        stdout=self.logfile, stderr=subprocess.PIPE)
+
+    logs.log('Waiting for emulated device to come online.')
+    while not new_device:
+      time.sleep(_WAIT_SECONDS)
+      for device in adb.get_devices():
+        if device not in devices_before:
+          environment.set_value('ANDROID_SERIAL', device)
+          new_device = True
+          logs.log('New device online with serial %s' % device)
+
+  def kill(self):
+    """ Kills the currently-running emulator, if there is one. """
+    if self.popen:
+      logs.log('Stopping emulator.')
+      self.popen.kill()
+      self.popen = None
+
+    if self.logfile:
+      self.logfile.close()
+      self.logfile = None
+
+    if self.emu_path:
+      shutil.rmtree(self.emu_path)
+
+
+def start_emulator():
+  """Start emulator."""
+  global _emu_proc
+  global _emu_users
+  if _emu_users == 0:
+    _emu_proc = EmulatorProcess()
+    _emu_proc.create()
+    _emu_proc.run()
+    adb.run_as_root()
+  _emu_users += 1
```

Comment:
Will we have more emu_users ? this logic seems to only account for one.

Reply:
Will there be more than one task running per bot? I assumed that we had 1 bot/task per process, and that we would thus be okay with a single emulator.

Reply:
not per bot, but there can be multiple cf bots per host right, wont they conflict with one another.

Reply:
why we need a counter on _emu_users when it can only have 0 or 1 ?

Reply:
The counter has been removed in the update that reads the child process's stdout.